src/data_generators/filesystem_path_client.py

```diff
@@ -0,0 +1,67 @@
+# Copyright 2021 The Private Cardinality Estimation Framework Authors
```

Comment:
Matthew, thank you for sharing your thought, and I am so sorry I didn't make it clear at the first place. I will write a one-page design doc to elaborate more details about the decision making, but here let me explain why I chose the current implementation and see if you would agree with me.

I've considered the design you drafted above, but then I found there are 2 drawbacks:
1. `(minor drawback)` With this implementation, the user can use either only local paths or only cloud paths in one request. There is no flexibility. 
2. `(major drawback)` Cloud clients initiated outside the Apache Beam pipeline can't be pickled and so can't be used inside the pipeline. Related discussion thread can be found [here](https://github.com/googleapis/google-cloud-python/issues/3191). This means the code block

    ```python
    if self.use_google_cloud():
        filesystem = GSFileSystem()
    else:
        filesystem = DefaultFileSystem()
    ```

    has to be used again in `ExperimentalTrial.evaluate` and potentially in other places when more functionalities use Apache Beam. If we are going to use a google-based filesystem for internal usage, this will make the code hard to maintain.

On the other hand, the implementation I have in this PR makes the decision in one place, `FilesystemPathClient`. In addition, it takes the advantage of the unity of the [filesystem path design](https://www.python.org/dev/peps/pep-0355/). `Pathlib`, `CloudPath`, `gpathlib`, or any other filesystem path classes, they all follow "almost" the same APIs, which reduces the need of a parent class.